ext/standard/var.c

```diff
@@ -815,36 +824,57 @@ static void php_var_serialize_get_sleep_props(
 		}
 
 		name = zval_get_tmp_string(name_val, &tmp_name);
-		if (php_var_serialize_try_add_sleep_prop(ht, props, name, name) == SUCCESS) {
+		if (php_var_serialize_try_add_sleep_prop(ht, props, name, name, struc) == SUCCESS) {
 			zend_tmp_string_release(tmp_name);
 			continue;
 		}
 
+		if (EG(exception)) {
+			zend_tmp_string_release(tmp_name);
+			retval = FAILURE;
+			goto exit;
+		}
+
 		priv_name = zend_mangle_property_name(
 			ZSTR_VAL(ce->name), ZSTR_LEN(ce->name),
 			ZSTR_VAL(name), ZSTR_LEN(name), ce->type & ZEND_INTERNAL_CLASS);
-		if (php_var_serialize_try_add_sleep_prop(ht, props, priv_name, name) == SUCCESS) {
+		if (php_var_serialize_try_add_sleep_prop(ht, props, priv_name, name, struc) == SUCCESS) {
 			zend_tmp_string_release(tmp_name);
 			zend_string_release(priv_name);
 			continue;
 		}
 		zend_string_release(priv_name);
 
+		if (EG(exception)) {
+			zend_tmp_string_release(tmp_name);
+			retval = FAILURE;
+			goto exit;
+		}
+
 		prot_name = zend_mangle_property_name(
 			"*", 1, ZSTR_VAL(name), ZSTR_LEN(name), ce->type & ZEND_INTERNAL_CLASS);
-		if (php_var_serialize_try_add_sleep_prop(ht, props, prot_name, name) == SUCCESS) {
+		if (php_var_serialize_try_add_sleep_prop(ht, props, prot_name, name, struc) == SUCCESS) {
 			zend_tmp_string_release(tmp_name);
 			zend_string_release(prot_name);
 			continue;
 		}
 		zend_string_release(prot_name);
 
+		if (EG(exception)) {
+			zend_tmp_string_release(tmp_name);
+			retval = FAILURE;
+			goto exit;
```

Comment:
It seems like `break;` might also work here, since there are no inner loops.

`goto exit` works better if there's ever a difference between successful exit after the loop, though.

```
#define ZEND_HASH_FOREACH_END() \
		} \
	} while (0)
```

Reply:
Good point! I've switch it to use break.

I generally hate the error handling inside this function, but don't see a great way to make it better without a larger refactoring (the TODO I added).